// @simple-obs/core/lib/node/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

const Scene = require('./Scene.js');

const Source = require('./Source.js');

const SceneItem = require('./SceneItem.js');

const Filter = require('./Filter.js');

const index = require('./obs/index.js');

require('./sources/index.js');

const index$1 = require('./filters/index.js');

const Image = require('./sources/Image.js');

const Browser = require('./sources/Browser.js');

const Color = require('./sources/Color.js');

const FreetypeText = require('./sources/FreetypeText.js');

const GDIPlusText = require('./sources/GDIPlusText.js');

const Media = require('./sources/Media.js');

const DisplayCapture = require('./sources/DisplayCapture.js');

const VideoCapture = require('./sources/VideoCapture.js');

exports.Scene = Scene.Scene;
exports.Source = Source.Source;
Object.defineProperty(exports, 'Alignment', {
  enumerable: true,
  get: () => SceneItem.Alignment
});
exports.DEFAULT_SCENE_ITEM_PROPERTIES = SceneItem.DEFAULT_SCENE_ITEM_PROPERTIES;
exports.SceneItem = SceneItem.SceneItem;
exports.Filter = Filter.Filter;
exports.obs = index.obs;
exports.ApplyLUTFilter = index$1.ApplyLUTFilter;
exports.ApplyLutFilter = index$1.ApplyLutFilter;
exports.AspectRatioFilter = index$1.AspectRatioFilter;
Object.defineProperty(exports, 'ChromaKeyColorType', {
  enumerable: true,
  get: () => index$1.ChromaKeyColorType
});
exports.ChromaKeyFilter = index$1.ChromaKeyFilter;
exports.ColorCorrectionFilter = index$1.ColorCorrectionFilter;
Object.defineProperty(exports, 'ColorKeyColorType', {
  enumerable: true,
  get: () => index$1.ColorKeyColorType
});
exports.ColorKeyFilter = index$1.ColorKeyFilter;
exports.CompressorFilter = index$1.CompressorFilter;
exports.CropPadFilter = index$1.CropPadFilter;
Object.defineProperty(exports, 'ExpanderDetectorType', {
  enumerable: true,
  get: () => index$1.ExpanderDetectorType
});
exports.ExpanderFilter = index$1.ExpanderFilter;
Object.defineProperty(exports, 'ExpanderPreset', {
  enumerable: true,
  get: () => index$1.ExpanderPreset
});
exports.GainFilter = index$1.GainFilter;
exports.ImageMaskBlendFilter = index$1.ImageMaskBlendFilter;
exports.InvertPolarityFilter = index$1.InvertPolarityFilter;
exports.LimiterFilter = index$1.LimiterFilter;
exports.LumaKeyFilter = index$1.LumaKeyFilter;
Object.defineProperty(exports, 'MaskBlendSelect', {
  enumerable: true,
  get: () => index$1.MaskBlendSelect
});
exports.NoiseGateFilter = index$1.NoiseGateFilter;
exports.NoiseSuppressFilter = index$1.NoiseSuppressFilter;
Object.defineProperty(exports, 'NoiseSuppressMethod', {
  enumerable: true,
  get: () => index$1.NoiseSuppressMethod
});
exports.RenderDelayFilter = index$1.RenderDelayFilter;
exports.ScrollFilter = index$1.ScrollFilter;
exports.SharpenFilter = index$1.SharpenFilter;
exports.ImageSource = Image.ImageSource;
exports.BrowserSource = Browser.BrowserSource;
exports.BrowserSourceItem = Browser.BrowserSourceItem;
exports.ColorSource = Color.ColorSource;
exports.FreetypeTextSource = FreetypeText.FreetypeTextSource;
exports.GDIPlusTextSource = GDIPlusText.GDIPlusTextSource;
exports.MediaSource = Media.MediaSource;
exports.DisplayCaptureSource = DisplayCapture.DisplayCaptureSource;
exports.VideoCaptureSource = VideoCapture.VideoCaptureSource;
